[ADD] connector_helpscout: Create HelpScout Conversation import

hughesbm opened the pull request:
Adds conversation import to HelpScout connector, including tags (as dependencies), threads, and attachments.

Depends on:
- [x] https://github.com/LasLabs/odoo-connector-helpscout/pull/7
- [x] https://github.com/LasLabs/python-helpscout/pull/26
- [x] https://github.com/LasLabs/python-helpscout/pull/35

hughesbm committed:
[ADD] connector_helpscout: Create HelpScout Conversation import

* Add importer, binding, adapter, mappings
* Add project_task_code to dependencies
* Add Conversation import to helpscout backend view
* Add mailbox folder to project.task form view
* Add tasks to helpscout.mailbox.folder form view
* Add project.task.type records (stages) for predefined helpscout
  statuses
* Add stages to helpscout mailboxes in after import hook
* Add Tag importing
  * Add importer, binding, adapter, mappings
  * Add tag import to conversation dependencies
  * Relate imported tags to conversation tasks
* Add Thread importing
  * Add importer, binding, adapter, mappings
  * Add thread import to conversation _after_import
  * Add message subtype records
* Add Attachment importing
  * Add importer, binding, adapter, mappings
  * Add attachment import to thread _after_import

lasley commented:
Is this ready for review?

hughesbm commented:
No, not yet. Importing is nearly ready (there are more changes that I haven't pushed up yet), but exporting still has a ways to go. Conversations/Threads/Attachments are all very interrelated so I've been working on them in parallel. Should I break the export portions out into a separate PR at this point?

lasley commented:
Yeah we'll do export in another PR in order to limit scope

hughesbm commented:
One feature that would be nice to have is marking the forward threads (thread type of `forwardchild` and `forwardparent` in HelpScout, which maps to mail.message subtypes `thread_forwardchild` and `thread_forwardparent`), somehow.

The `lineitem` threads (used in HelpScout for internal notes like status changes, etc.) aren't updatable on HelpScout as far as I can tell, so for those I'm simply adding a useful description of the action to the otherwise blank message body in Odoo. But the `forwardchild` / `forwardparent` threads _can_ be updated on HelpScout, so we would have to do some messy string parsing to get rid of any additions to the body during export, if we wanted to use the same strategy.

I was looking at using the `mail.message.subtype` description field as an alternative, but I'm not sure if it's the right place, as it seems to have something to do with tracking. Not quite sure how that field is supposed to be used.

lasley commented on connector_helpscout/models/helpscout_backend/common.py:
```diff
@@ -24,12 +24,15 @@ class HelpscoutBackend(models.Model):
     _inherit = 'connector.backend'
 
     EVENT_TO_MODEL = {
+        'attachment': 'helpscout.attachment',
```
This is meant to map web hook prefixes to models they reference. There aren't `attachment`, `tag`, and `thread` hooks though - what are you using this for?

hughesbm replied:
Whoops. I didn't look up the events, just kept adding as I went.

lasley commented on connector_helpscout/models/helpscout_backend/common.py:
```diff
@@ -73,6 +76,7 @@ class HelpscoutBackend(models.Model):
     )
 
     # These dates are used by the import crons to start where left off.
+    import_conversations_from_date = fields.Datetime()
```
We don't need this, we're going to use web hooks to receive changes (`action_receive_hook`). We only need the ability to do a mass import once for all data, which can be handled with `binding_model.import_batch(backend)`

lasley commented on connector_helpscout/models/helpscout_backend/common.py:
```diff
@@ -194,17 +198,24 @@ def _get_hook_values(self):
                               'import_customers_from_date')
 
+    @api.multi
+    def action_import_mailboxes(self):
```
We'll just implicitly import the mailboxes as dependencies to conversations, as well as provide an initial import hook. No need for cron method

lasley commented on connector_helpscout/models/helpscout_conversation/importer.py:
```diff
@@ -0,0 +1,131 @@
+        folder_binding = self.env['helpscout.helpscout.folder'].search([
+            ('external_id', '=', folder_external_id)
+        ])
+        folder = folder_binding.odoo_id
```
Kill this line and do an if on the binding directly, then use directly

lasley commented on connector_helpscout/models/helpscout_conversation/importer.py:
```diff
@@ -0,0 +1,131 @@
+            ('name', 'in', record.tags)
+        ])
+        if tags:
+            tag_ids = tags.mapped('id')
```
Kill this line and just use `tags.ids` below

lasley commented on connector_helpscout/models/helpscout_conversation/importer.py:
```diff
@@ -0,0 +1,131 @@
+    @mapping
+    def tag_ids(self, record):
+        tags = self.env['project.tags'].search([
+            ('name', 'in', record.tags)
```
trailing comma

lasley commented on connector_helpscout/models/helpscout_conversation/importer.py:
```diff
@@ -0,0 +1,131 @@
+    def user_id(self, record):
+        try:
+            owner = self.env['helpscout.user'].search([
+                ('external_id', '=', record.owner.id)
```
trailing comma

lasley commented on connector_helpscout/models/helpscout_conversation/importer.py:
```diff
@@ -0,0 +1,131 @@
+            tracking_disable=True,
+        )
+        thread_ids = [
+            binding_model.import_direct(self.backend_record, thread).id
```
You should always just work with recordsets instead of mapping to `ids`. So this would be `threads = [..]` & the below `.browse` should be removed. Or was this for a specific reason?

hughesbm replied:
The idea here was to get a recordset for the res_id/model write, since I just have a list of newly-imported thread records at this point. Is there a better way to do this?

lasley replied:
Well I think you're doing this fine, it's just that you should ALWAYS be working with recordsets until the last possible moment. In this case, you're creating a list of IDs that could easily be obtained with `recordset.ids`

hughesbm replied:
But I don't have a recordset until I get it with `.browse`, just a list of individual records that were just imported. Or am I missing something obvious?

lasley replied:
AFAIK the binding record is returned from `importer.run`, which is returned by `import_direct` & alluded to here by the fact that you are performing a `.id` on the result:
```
binding_model.import_direct(self.backend_record, thread).id
```

So instead I am recommending:

```
thread_bindings = sum(
    binding_model.import_direct(self.backend_record, thread)
    for thread in self.helpscout_record['threads']
)
```

or something like that (questionable whether `sum` will work as intended there). Then:

```
thread_bindings.write({
    'model': ...
```

hughesbm replied:
Ah, thanks. I follow you on working with recordsets when possible, I was just trying to get something I could actually call `.write` on. I'll try `sum`.

lasley replied:
if sum doesn't work, just do like this:

```
thread_bindings = binding_model
for thread in self.helpscout_record['threads']:
    thread_bindings += import
```

lasley commented on connector_helpscout/models/helpscout_conversation/importer.py:
```diff
@@ -0,0 +1,131 @@
+            self.backend_record,
+            {'mailbox_id': self.helpscout_record['mailbox']['id']},
+        )
+        self.env['helpscout.tag'].import_batch(self.backend_record)
```
Hmmm now that I think about it, instead of doing this every time (tags probably won't change much), we can save a lot of API calls by doing it in the tag_ids mapping method instead.

We should search for pre-existing tags, then compare against the amount of tags that actually exist in the record. If they do not match, call `import_batch`, then return the recursive value:

```python
@mapping
def tag_ids(self, record):
    tags = self.env['helpscout.project.tags'].search([..])
    if len(tags) != len(record.tags):
        self.env['helpscout.project.tags'].import_batch(backend)
        return self.tag_ids(record)
    return {'tag_ids': [(6, 0, tags.ids)]}
```

lasley commented on connector_helpscout/models/helpscout_customer/importer.py:
```diff
@@ -26,6 +26,15 @@ def email(self, record):
             return
 
+    @mapping
+    def full_name(self, record):
```
Weird. When does this happen?

hughesbm replied:
The specific case that led to this was the helpscout postmaster, but I assumed if that got through, there may be others.

lasley commented on connector_helpscout/models/helpscout_conversation/common.py:
```diff
@@ -0,0 +1,54 @@
+        comodel_name='helpscout.conversation',
+        inverse_name='odoo_id',
+    )
+    helpscout_folder_id = fields.Many2one(
```
Hmmm typically we want all helpscout context in the bindings, vs having them in the concrete models. Is there a reason for this?

lasley commented on connector_helpscout/models/helpscout_folder/common.py:
```diff
@@ -30,6 +30,11 @@ class HelpScoutFolder(models.Model):
         string='HelpScout Folder Name',
         required=True,
     )
+    task_ids = fields.One2many(
```
Typically in bind models, we will want to join with other bind models. This keeps a consistency that is otherwise really hard to mentally picture.

lasley commented on connector_helpscout/models/helpscout_mailbox/importer.py:
```diff
@@ -34,6 +34,15 @@ class HelpScoutMailboxImporter(Component):
+        """Add HelpScout status choices as project stages"""
+        helpscout_statuses = ['active', 'pending', 'closed', 'spam']
+        stages = [self._get_stage_id(status) for status in helpscout_statuses]
+        binding.type_ids = [(6, 0, stages)]
```
`stages.ids` probably

lasley commented on connector_helpscout/models/helpscout_mailbox/importer.py:
```diff
@@ -34,6 +34,15 @@ class HelpScoutMailboxImporter(Component):
+        stages = [self._get_stage_id(status) for status in helpscout_statuses]
+        binding.type_ids = [(6, 0, stages)]
+
+    def _get_stage_id(self, status):
```
Always work with recordsets. Never ids until the last possible moment

lasley commented on connector_helpscout/models/helpscout_tag/common.py:
```diff
@@ -0,0 +1,54 @@
+    def search(self, filters=None):
+        return [r.id for r in self.search_read()]
+
+    def read(self, _id):
```
`external_id` instead of `_id`. The underscore prefix in an arg is basically saying you're not going to use it, but are defining it anyways (typically for inheritance). It will circumvent a lint error for declaring an arg that's not being used.

lasley commented on connector_helpscout/models/helpscout_tag/importer.py:
```diff
@@ -0,0 +1,61 @@
+
+
+tag_color_map = {
+    (211, 211, 211): 0,
```
Hrmmm maybe some inline comments to map these colors for human consumption too?

lasley commented on connector_helpscout/models/helpscout_tag/importer.py:
```diff
@@ -0,0 +1,61 @@
+
+    direct = [('tag', 'name'),
+              ('slug', 'helpscout_slug'),
+              ('created_at', 'backend_date_created'),
```
This is already done in the default mapper https://github.com/LasLabs/odoo-connector-helpscout/blob/10.0/connector_helpscout/components/mapper.py

lasley commented on connector_helpscout/models/helpscout_thread/common.py:
```diff
@@ -0,0 +1,119 @@
+        return self.endpoint.create_thread(conversation_id,
+                                           self.new_record(data))
+
+    def write(self, _id, data, conversation_id):
```
Hmmm you're going to have a hard time injecting the extra arg here I think. Is the conversation ID not already encapsulated within the data? I feel like it should be.

hughesbm replied:
It's not in the thread object, unfortunately (https://developer.helpscout.com/help-desk-api/objects/thread/).

These methods are also out of scope since this PR doesn't include exporting, so I'll remove them for now.

lasley reviewed (changes requested):
Still a few files left to review

hughesbm committed:
[IMP] Make requested changes

* Remove references to non-existent events
* Move helpscout relations from concrete models to bindings
* Add missing trailing commas
* Use simplified tags.ids mapping
* Refactor Conversation, Mailbox _after_import methods
* Replace _id with external_id in Tag adapter read method
* Add tag color inline comments
* Remove created_at, modified_at direct mappings from importers
* Remove out-of-scope Thread adapter methods
* Import tags during Conversation tag_ids import mapping if tags
  are missing, rather than as dependency
* Import Mailboxes as Conversation dependency
* Simplify bulk imports
  * Remove cron actions, datetime fields, date filtering
  * Update backend view
* Clean up Folder import
  * Import Folders in Mailbox _after_import instead of as
    Conversation dependency
  * Remove messy overrides of external_id field, adapter methods
* Add Customer, User as Conversation dependencies
* Add access control records

hughesbm commented:
The requested changes have been made, as well as some additional improvements, including:

- Simplified bulk import (no more crons)
- Added Customer, User as Conversation import dependencies (things get ugly if these are missing, especially during export)
- Simplified folder import (similar to how threads, attachments are imported)
- Added missing access control records

lasley commented:
CI is failing:

> UserError: (u"You try to install module 'connector_helpscout' that depends on module 'project_task_code'.\nBut the latter module is not available in your system.", '')

To fix, you just need to add `project` in `oca_dependencies.txt`

hughesbm committed:
[FIX] Add project to oca_dependencies

lasley commented on connector_helpscout/models/helpscout_conversation/importer.py:
```diff
@@ -70,18 +62,18 @@ def stage_id(self, record):
-        tags = self.env['project.tags'].search([
-            ('name', 'in', record.tags)
-        ])
+        tags = self.env['project.tags'].search([('name', 'in', record.tags)])
```
You should probably search on `helpscout.project.tags`, and add `backend_id` to the domain. Otherwise all tags are included, which will result in an infinite recursion if a tag is added from outside of this connector.

lasley commented on connector_helpscout/models/helpscout_mailbox/importer.py:
```diff
@@ -35,13 +33,28 @@ class HelpScoutMailboxImporter(Component):
 
     def _after_import(self, binding):
-        """Add HelpScout status choices as project stages"""
+        """
```
Blank space is against [PEP-0257](https://www.python.org/dev/peps/pep-0257/#multi-line-docstrings)

lasley commented on connector_helpscout/models/helpscout_mailbox/importer.py:
```diff
@@ -35,13 +33,28 @@ class HelpScoutMailboxImporter(Component):
+        self._import_folders(binding)
+
+    def _add_project_stages(self, binding):
         helpscout_statuses = ['active', 'pending', 'closed', 'spam']
```
Might make sense to make this a class level constant

lasley commented on connector_helpscout/models/helpscout_tag/importer.py:
```diff
@@ -23,17 +23,13 @@ class HelpScoutTagImportMapper(Component):
-        tag = self.env['project.tags'].search([
-            ('name', '=', record.tag),
-        ])
+        tag = self.env['project.tags'].search([('name', '=', record.tag)])
```
Do we know if name is unique? If not, either limit search to one or do `tag[:1].id` in the return. The limit is better though because less data is being tossed around.

lasley commented on connector_helpscout/views/helpscout_backend_view.xml:
```diff
@@ -11,6 +11,11 @@
+                    <button name="action_bulk_import"
+                            type="object"
+                            class="oe_highlight"
+                            string="Queue Bulk Import"
```
I'd probably call this `Queue Initial Import` instead of bulk.

lasley reviewed (changes requested):
Thanks @hughesbm

hughesbm committed:
[IMP] Make requested changes

* Improve tag search in Conversation tag_ids mapping
* Remove unhelpful docstring
* Move helpscout statuses to class-level constant
* Limit tag search in Tag odoo_id mapping to one record
* Rename Queue Bulk Import -> Queue Initial Import

lasley reviewed (approved):
Thanks @hughesbm
